File: examples/answer-trace/apps/api/app/services/cf.py
"""LLM 客户端 —— OpenAI SDK 指向任意 OpenAI 兼容端点。

base_url / api_key 见 config:默认 Cloudflare Workers AI,也可设 AT_BASE_URL +
AT_API_KEY 切到 OpenRouter / DeepSeek / z.ai 等(模型 ID 也对应改 AT_GEN_MODEL/AT_JUDGE_MODEL)。
"""

import re
from functools import lru_cache
from typing import Any

from openai import OpenAI

from app import config

# qwen3 这类带 thinking 的模型会先吐 <think>…</think>,会污染 JSON,统一剥掉。
_THINK = re.compile(r"<think>.*?</think>", re.DOTALL)


@lru_cache(maxsize=1)
def _client() -> OpenAI:
    return OpenAI(base_url=config.base_url(), api_key=config.api_key())


def chat(
    model: str,
    messages: list[dict[str, Any]],
    *,
    temperature: float = 0.0,
    json_mode: bool = False,
) -> str:
    """调用一次 chat,返回纯文本(已剥离 thinking)。json_mode 尽力要求 JSON 输出。"""
    kwargs: dict[str, Any] = {"model": model, "messages": messages, "temperature": temperature}
    if json_mode:
        kwargs["response_format"] = {"type": "json_object"}

    try:
        resp = _client().chat.completions.create(**kwargs)
    except Exception:
        # 个别模型/端点不接受 response_format,去掉重试一次(靠 prompt 约束 JSON)。
        if json_mode:
            kwargs.pop("response_format", None)
            resp = _client().chat.completions.create(**kwargs)
        else:
            raise

    text = resp.choices[0].message.content or ""
    return _THINK.sub("", text).strip()

File: examples/answer-trace/apps/edge/src/cf.ts
// LLM 客户端：fetch 任意 OpenAI 兼容端点（/chat/completions）。
// 等价于 apps/api/app/services/cf.py（那边用 openai SDK，这里直接 fetch，零依赖）。

import { cfg, type Env } from "./config";

// qwen3 这类带 thinking 的模型会先吐 <think>…</think>，污染 JSON，统一剥掉。
const THINK = /<think>[\s\S]*?<\/think>/g;

export interface ChatMessage {
  role: "system" | "user" | "assistant";
  content: string;
}

interface ChatOptions {
  temperature?: number;
  jsonMode?: boolean;
}

interface ChatBody {
  model: string;
  messages: ChatMessage[];
  temperature: number;
  response_format?: { type: "json_object" };
}

async function post(url: string, apiKey: string, body: ChatBody): Promise<Response> {
  return fetch(url, {
    method: "POST",
    headers: {
      authorization: `Bearer ${apiKey}`,
      "content-type": "application/json",
    },
    body: JSON.stringify(body),
  });
}

/** 调用一次 chat，返回纯文本（已剥离 thinking）。jsonMode 尽力要求 JSON 输出。 */
export async function chat(
  env: Env,
  model: string,
  messages: ChatMessage[],
  opts: ChatOptions = {},
): Promise<string> {
  const c = cfg(env);
  const url = `${c.baseUrl.replace(/\/$/, "")}/chat/completions`;
  const body: ChatBody = {
    model,
    messages,
    temperature: opts.temperature ?? 0,
  };
  if (opts.jsonMode) body.response_format = { type: "json_object" };

  let resp = await post(url, c.apiKey, body);
  if (!resp.ok && opts.jsonMode) {
    // 个别端点不接受 response_format，去掉重试一次（靠 prompt 约束 JSON）。
    const { response_format: _omit, ...rest } = body;
    resp = await post(url, c.apiKey, rest as ChatBody);
  }
  if (!resp.ok) {
    throw new Error(`模型端点返回 ${resp.status}：${(await resp.text()).slice(0, 300)}`);
  }

  const data = (await resp.json()) as { choices?: { message?: { content?: string } }[] };
  const text = data.choices?.[0]?.message?.content ?? "";
  return text.replace(THINK, "").trim();
}
